### PR TITLE
Automate review app building via CircleCI

### PR DESCRIPTION
**Pull request by dleve123**
Problem

Review apps are great, particularly for long-running work like redesigns, major refactors and anything else that might risk causing regressions (blocking our deploy pipeline), but need further testing in a live environment.

During a platform practice meeting, it was surfaced that:

1. It's a bit too hard to create/update a review app, even with build_review_app.sh
2. It can be **super slow** to build and push a Docker image of Force (~2 GB) depending on the conditions of the developer host running build_review_app.sh and the network.

Solution

* Create a review app when pushing to a branch matching review-app-*: d0d7edf82b83230ab741296a1404f1f104d98090
* Update a review app if it already exists upon subsequent commits: 6cb75b7aeb35cf74e963337e4396ef17391c6cdc
* Remove data dog deploy instrumentation after a Hokusai deploy: bdc804f380b8d9afb785c09c9deaf6973d2be314

Impact

It now takes ~15 minutes, consistently to create/update a Force review app, similar to the time takes to deploy `master` to `staging`

TODO

- [x] Remove commented out build steps / rebase away 82f1373
- [x] Detect existence of existing review app and update it instead of creating

### CODE REVIEW

**Comment by ArtsyOpenSource** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>The staging version of Metaphysics has breaking changes, this might not affect your build, but could mean your queries don't work the way you expect. See below for breaking changes:</td>
    </tr>
  </tbody>
</table>


- :warning: Field 'artworksByInternalID' was removed from object type 'Query'
- :warning: Field 'artworksByInternalID' was removed from object type 'Viewer'
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 87c1695cc0f0124e2c85a9e11be9f6d5cfe025ce
</p>

**Comment by dleve123** [comment]
@izakp Looks like we're running into a permissioning issue with the `kubernetes-ci` user. Looks like it can't create namespaces:

```
# Provision the review app
hokusai review_app create $NAME --verbose
Error from server (Forbidden): error when creating "/root/project/hokusai/building-on-circle.yml": namespaces is forbidden: User "kubernetes-ci" cannot create namespaces at the cluster scope
Error from server (NotFound): error when creating "/root/project/hokusai/building-on-circle.yml": namespaces "building-on-circle" not found
Error from server (NotFound): error when creating "/root/project/hokusai/building-on-circle.yml": namespaces "building-on-circle" not found
Error from server (NotFound): error when creating "/root/project/hokusai/building-on-circle.yml": namespaces "building-on-circle" not found
ERROR: Command 'kubectl --context staging --namespace building-on-circle create --save-config -f /root/project/hokusai/building-on-circle.yml' returned non-zero exit status 1
```

(https://app.circleci.com/pipelines/github/artsy/force/5821/workflows/f73e0681-b889-4dda-9048-0d1351efefff/jobs/42771/parallel-runs/0/steps/0-107)

Would it be possible to allow this user to create namespaces? Thoughts?

**Comment by dleve123** [comment]
cc @damassi for his context

**Comment by dblandin** [comment]
> Would it be possible to allow this user to create namespaces? Thoughts?

When we last chatted about this, we were thinking of adjusting how review apps are created so that they all share a common "review-apps" namespace. If that were the case, the kubernetes-ci user could be given access to manage that specific namespace.

In the short-term, we could consider broader namespace access to get over this hurdle.

**Comment by dleve123** [comment]
So cool, this built a Force review app at https://a47e161eb79c611ea976f123b14b93be-1668184118.us-east-1.elb.amazonaws.com/!

Screen Shot 2020-04-10 at 4 53 40 PM

 Gonna fix the `jq` error and delete the namespace to see if we can get a totally fresh build

File: .circleci/config.yml [hunk 1]
```diff
@@ -56,6 +56,19 @@ jobs:
...
 
+  build_review_app:
+    executor: hokusai/deploy
+    steps:
+      - hokusai/setup-docker
+      - hokusai/install-aws-iam-authenticator
+      - hokusai/configure-hokusai
+      - run:
+          name: Install jq
+          command: curl -L https://github.com/stedolan/jq/releases/download/jq-1.6/jq-linux64 > /usr/local/bin/jq
```
Comment (dblandin) [inline .circleci/config.yml:67]: We could also consider installing jq within the hokusai Docker image: https://github.com/artsy/hokusai/blob/master/Dockerfile

**Comment by dleve123** [comment]
next blocker:

Looks like something about the upgrade to using jinja vars in hokusai/staging.yml broke `hokusai review_app setup`. I'm able to replicate the error on the circle build locally:


```dl@ip-192-168-1-226:~/src/force (master)
$ hokusai review_app setup foobar
Cannot read source yaml file hokusai/staging.yml.
dl@ip-192-168-1-226:~/src/force (master)
$ git checkout 3ac469c1f33bb3d1c767b9e5d65cfef4268e094a -- hokusai/staging.yml
dl@ip-192-168-1-226:~/src/force * (master)
$ git st
On branch master
Your branch is up to date with 'upstream/master'.

Changes to be committed:
  (use "git reset HEAD <file>..." to unstage)

	modified:   hokusai/staging.yml

dl@ip-192-168-1-226:~/src/force * (master)
$ git diff
dl@ip-192-168-1-226:~/src/force * (master)
$ git diff --staged
diff --git a/hokusai/staging.yml b/hokusai/staging.yml
index b504dd813..79586db9f 100644
--- a/hokusai/staging.yml
+++ b/hokusai/staging.yml
@@ -121,7 +121,7 @@ metadata:
   name: force-web
   namespace: default
   annotations:
-    service.beta.kubernetes.io/aws-load-balancer-ssl-cert: {{ artsyNetWildcardSSLCert }}
+    service.beta.kubernetes.io/aws-load-balancer-ssl-cert: "arn:aws:iam::585031190124:server-certificate/2018-01-17_artsy-net-wildcard"
     service.beta.kubernetes.io/aws-load-balancer-backend-protocol: "http"
     service.beta.kubernetes.io/aws-load-balancer-ssl-ports: "443"
     service.beta.kubernetes.io/aws-load-balancer-cross-zone-load-balancing-enabled: "true"
dl@ip-192-168-1-226:~/src/force * (master)
$ hokusai review_app setup foobar
Created hokusai/foobar.yml
dl@ip-192-168-1-226:~/src/force * (master)
$ hokusai version
0.5.8
dl@ip-192-168-1-226:~/src/force * (master)
$ git status
On branch master
Your branch is up to date with 'upstream/master'.

Changes to be committed:
  (use "git reset HEAD <file>..." to unstage)

	modified:   hokusai/staging.yml

Untracked files:
  (use "git add <file>..." to include in what will be committed)

	hokusai/foobar.yml

dl@ip-192-168-1-226:~/src/force * (master)
```

TL;DR of the above – soft reverting https://github.com/artsy/force/commit/b25dad646090a2ca7b12bbcc5a87267f6c9908a1#diff-9af8dd47d98e3e7912dadcf50c599d06 results in the review app being set-up able again. Seems like a side-effect of the 0.5.8 upgrade?

cc @eessex @izakp

**Comment by dleve123** [comment]
16 minutes to update an existing review app on Circle:

> Duration 16m 20s

https://app.circleci.com/pipelines/github/artsy/force/6320/workflows/fb66b446-66b8-448e-99c6-3e7d4b2e628c/jobs/45512/steps

Gonna delete the review app name and then hopefully time how long it takes to build from scratch.

**Comment by dleve123** [comment]
This is ready for review! Naturally, there is some build flakiness due to I think a Github availability issue (we've seen it before), but previous comments on this PR attest to this working (with links to CircleCI builds).

Assigned myself so that I can handle the rebase and merge, but would love 2+ approved reviews!

File: scripts/update_review_app.sh [hunk 2]
```diff
@@ -0,0 +1,18 @@
+#!/bin/bash
+
+set -ev
+
+# Bail out of script on first expression failure and echo the commands as
+# they are being run.
+set -ev
```
Comment (dblandin) [inline scripts/update_review_app.sh:7]: Looks like this is duplicated on L3.

File: scripts/update_review_app.sh [hunk 3]
```diff
@@ -0,0 +1,18 @@
+#!/bin/bash
```
Comment (dblandin) [inline scripts/update_review_app.sh:1]: `#!/usr/bin/env bash` is usually more portable than expecting a specific system path.

File: scripts/update_review_app.sh [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+
+set -ev
+
+# Bail out of script on first expression failure and echo the commands as
+# they are being run.
+set -ev
+
+NAME="$1"
+
+if test -z $NAME; then
```
Comment (dblandin) [inline scripts/update_review_app.sh:11]: We should wrap any variable in double quotes to prevent globbing and word splitting.

File: scripts/build_review_app.sh [hunk 5]
```diff
@@ -34,29 +35,24 @@ hokusai review_app setup $NAME
...
+hokusai registry push --force --skip-latest --overwrite --verbose --tag $NAME
 
 # Edit the K8S YAML to reference the proper Docker image
-# sed -i '' 's/:staging/:auct-281/g' hokusai/auct-281.yml
-#
-# The weird `-i ''` is needed on MacOS:
-# https://stackoverflow.com/a/44864004
-# WARNING: This might not work on Linux.
-sed -i '' "s/:staging/:$NAME/g" hokusai/$NAME.yml
+sed -i.bak "s/:staging/:$NAME/g" $review_app_file_path && rm $review_app_file_path.bak
```
Comment (dblandin) [inline scripts/build_review_app.sh:41]: We should wrap any variables in double quotes to prevent globbing and word splitting.

File: scripts/build_review_app.sh [hunk 6]
```diff
@@ -34,29 +35,24 @@ hokusai review_app setup $NAME
...
 
 # Edit the K8S YAML to remove the instructions that enforce that the service
 # can only be accessible via Cloudflare
 #
 # First, remove the `loadBalancerSourceRanges:` line
-sed -i '' '/loadBalancer/d' hokusai/$NAME.yml
+sed -i.bak '/loadBalancer/d' $review_app_file_path && rm $review_app_file_path.bak
```
Comment (dblandin) [inline scripts/build_review_app.sh:47]: Are we expecting to be able to leverage the recently introduced Jinja template support in hokusai to clean up some of these sed/regex string replacements? Perhaps in a follow-up iteration?

File: scripts/build_review_app.sh [hunk 7]
```diff
@@ -34,29 +35,24 @@ hokusai review_app setup $NAME
 # --tag to name the image.
 # WARNING: This is likely going to take ~10 mins on your MBP.
 # Be patient and grab some baby carrots.
-hokusai registry push --force --skip-latest --verbose --tag $NAME
+hokusai registry push --force --skip-latest --overwrite --verbose --tag $NAME
```
Comment (dblandin) [inline scripts/build_review_app.sh:38]: `--overwrite` makes sense for `update_review_app.sh` but do we expect to run this build script more than once for a particular review app?

Nit: We might want to drop the file extension or update to `.bash` as we're specifying `bash` not `sh`as the target program.

File: scripts/update_review_app.sh [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+  echo "You didn't provide a shell argument, so NAME isn't meaningful, exiting."
+  exit 1
+fi
+
+hokusai registry push --force --skip-latest --overwrite --verbose --tag $NAME
+hokusai review_app setup $NAME
```
Comment (anandaroop) [inline scripts/update_review_app.sh:17]: Is the `setup` step necessary for an existing review app?

**Comment by dleve123** [comment]
**Security**: I wanted to make sure that this didn't expose any staging infra vectors into our cluster. I opened up this PR as evidence that it does not:

https://github.com/artsy/force/pull/5535

Note: Circle seems to set the `branch` of a forked PR to `pull/#{pr_number}`, so no branch triggers get fired:

https://app.circleci.com/pipelines/github/artsy/force/6342/workflows/16384db8-99eb-434b-b2c1-8c3df39225dd

With that said, I might have not exactly replicated this vector and would love some review on that. @joeyAghion as the defacto security DRI, can I have your sign off here?

**Comment by damassi** [comment]
@dleve123 — i might be a dense here, but how does one see the review app that they created? Noticed there’s no mention of a "check out x url" step anywhere in description above.

**Comment by dleve123** [comment]
> Did you consider using a label instead? Haven't thought about it much but that would free this functionality to work on forked PRs and as an after-thought to existing branches. (Now that I think about it, maybe that's too dangerous.)

We discussed labels, but wanted to get any working automation first. And yeah, def don't want to build on forks (see my previous comment). I think having a push for a commit trigger a build results in a nice and easy to see correlation between source code revision <-> docker image <-> running review app as well. Easy to connect the dots. 

> A follow-up question to this should be: how can we automatically clean up these review apps. If this is successful (and I expect it will be), review apps will be easier to use and we'll tend to have more of them. We're already running into some limitations (that are fixable but still hurdles) like excessive server resources needed in staging and default ELB limits.

I think @izakp discussed some form of a cron job that maybe deletes namspaces (in the staging cluster that haven't been updated) within N number of days). I believe there's an operations cluster that could be used for this? Happy to help think through this!

**Comment by dleve123** [comment]
> @dleve123 — i might be a bit dense here, but how does one see the review app that they created? Noticed there’s no mention of a "check out x url" step in description above.

@damassi not dense at all and you raise a good point. When `create_or_update_review_app` runs so that a review app is created, the last line of the bash script uses `jq` and interrogates our k8s cluster to output the external hostname of the new review app:

```
# yeah, I'm talking to you
 
# you need to configure a CNAME record so that $NAME.artsy.net resolves to" \
 
# the service's external IP:
 
#
 
echo $(kubectl get service force-web --namespace $NAME --context staging -o json \
 
  | jq .status.loadBalancer.ingress[].hostname)
 
"aa7f919f090af11ea976f123b14b93be-560250530.us-east-1.elb.amazonaws.com"
 
#
 
# you may do this in the Cloudflare interface. Credentials are in 1pass
 
echo "[build_review_app.sh] SUCCESS"
 
[build_review_app.sh] SUCCESS
 ```

(https://app.circleci.com/pipelines/github/artsy/force/6322/workflows/b81f0c8a-8175-4686-9996-6d933b346fe9/jobs/45522/steps)

It's def a bit burried though, would be cool to have that string surfaced directly in the build. Seems like CircleCI isn't quite yet there yet, but this is something that perhaps we can work into Peril or a Github action? Either way good q.

File: scripts/update_review_app.sh (hunk 8 above)
Comment (dleve123) [reply scripts/update_review_app.sh:17]: Weirdly, it is. I tried to elaborate in the commit message for https://github.com/artsy/force/pull/5370/commits/116e3e01ca810fc3d0cb0a14be40ec3e8dfb7b4e 

```
* Create a review app YAML file before deploying it.
  + Because review app management in rooted in a file on the FS, we need to
    have a "stub" review app YAML file to effectively re-deploy code to a
    review app.
```

File: scripts/build_review_app.sh (hunk 6 above)
Comment (dleve123) [reply scripts/build_review_app.sh:47]: > Are we expecting to be able to leverage the recently introduced Jinja template support in hokusai to clean up some of these sed/regex string replacements? Perhaps in a follow-up iteration?

I'm not aware of any plans to further parameterize these IPs, but that does seem like a good use-case and possible refactoring of the core hokusai setup here cc @izakp

File: scripts/build_review_app.sh (hunk 5 above)
Comment (dleve123) [reply scripts/build_review_app.sh:41]: `$NAME` is wrapped in double quotes earlier in the script: https://github.com/artsy/force/blob/d3bc6b4d07dcecc30d9869be1d4977d64ce64482/scripts/build_review_app.sh#L17
and `$review_app_file_path` is as well: https://github.com/artsy/force/blob/d3bc6b4d07dcecc30d9869be1d4977d64ce64482/scripts/build_review_app.sh#L17

As such, I think this is fine (aka there's no added risk here)?

File: scripts/update_review_app.sh (hunk 4 above)
Comment (dleve123) [reply scripts/update_review_app.sh:11]: Check out L9 above – we're wrapping `$1` in double quotes, so I think that's sufficient?

**Comment by damassi** [comment]
I see 👍 So in that case, mind updating the shell-script comments that we see in the CI output to instead point at something canonical like our [review app docs](https://github.com/artsy/force/blob/master/docs/creating_review_app.md#if-your-unmerged-changes-are-in-force)? Being less familiar with the devops side of things I've gotten hung up on last steps a few times. If the script pointed at these docs, then we would only have to point at one spot as things change through time.

File: scripts/update_review_app.sh (hunk 2 above)
Comment (dleve123) [reply scripts/update_review_app.sh:7]: Whoops - good catch. Thanks 👍

File: scripts/build_review_app.sh (hunk 5 above)
Comment (dblandin) [reply scripts/build_review_app.sh:41]: > As such, I think this is fine (aka there's no added risk here)?

Unfortunately not.

ShellCheck is a great tool for picking up these issues: https://github.com/koalaman/shellcheck

Here's an a simple example:

```bash
$ touch "an example.txt"
$ file="an example.txt"
$ rm 'some example.txt' ^C
$ ls $file
ls: cannot access 'an': No such file or directory
ls: cannot access 'example.txt': No such file or directory
$ ls "$file"
'an example.txt'
```

So even though you've wrapped $1 in double quotes, you should continue to do so for variables passed into other programs.

Comment (dleve123) [reply scripts/build_review_app.sh:41]: hu interesting! Well TIL. Yeah, I was thinking about it slightly off I guess (values of the vars can't be "escaped once" or anything like that), this makes sense. 👍

File: scripts/update_review_app.sh (hunk 3 above)
Comment (dleve123) [reply scripts/update_review_app.sh:1]: Going to pass on this for now. Did some quick reading and realized it seems like there are pros and cons on both sides here. At this point, the goal of the script isn't to be portable across systems, but support the specific Docker container running in Circle

File: scripts/build_review_app.sh (hunk 7 above)
Comment (dleve123) [reply scripts/build_review_app.sh:38]: > --overwrite makes sense for update_review_app.sh but do we expect to run this build script more than once for a particular review app?

While it's not expected, I do think this makes the script a bit more robust: E.g – for whatever reason, but a script run fails after this command, but before review app namespace creation. But allowing us to override here, a re-build is possible. Otherwise, we would have to manually delete the image in the ECR registry to proceed (without modifying the script).

> Nit: We might want to drop the file extension or update to .bash as we're specifying bash not shas the target program.

Personally I've never seen this done before for many a bash script. Gonna opt to not rename for now.

**Comment by dleve123** [comment]
@damassi @anandaroop @dblandin – I believe I addressed all your comments! Gonna merge in a couple of hours unless I hear from ya otherwise.

@damassi Also took the liberty to add to/update the review app docs in https://github.com/artsy/force/pull/5370/commits/6c513b30220bb5d90115dbed04bd7fef61d252b6

**Comment by damassi** [comment]
Looks great -- thanks again for all your work on this! I feel like the work carved out here around the branching pattern may set us up well for future automation ideas 👍
